pkgs/by-name/la/lammps/package.nix

```diff
@@ -44,9 +45,22 @@
   extraCmakeFlags ? { },
   # Extra `buildInputs` - meant for packages that require more inputs
   extraBuildInputs ? [ ],
+  # If false, the `stdenv` argument will be used no matter what is the state of
+  # extraCmakeFlags.GPU_API or config.cudaSupport . See logic for stdenv' below.
+  overrideStdenv ? true,
+  config ? { },
 }:
 
-stdenv.mkDerivation (finalAttrs: {
+let
+  # Ease adding CUDA support. See:
+  # https://github.com/lammps/lammps/issues/4420
+  stdenv' =
+    if overrideStdenv && (config.cudaSupport || ((extraCmakeFlags.GPU_API or "") == "cuda")) then
```

Comment:
In case the "cudaSupport" is true do you want to also set the corresponding `cmakeFlags` and `buildInputs` so that lammps gets build automatically with CUDA (i.e. in a similar way we do it in openmpi)?

Reply:
> In case the "cudaSupport" is true do you want to also set the corresponding `cmakeFlags` and `buildInputs` so that lammps gets build automatically with CUDA (i.e. in a similar way we do it in openmpi)?

I think it makes the logic pretty complicated and I'm not sure I want to delve into implementing this. How many details do we want to be taken care of automatically? What do we want to happen in case someone simply sets `config.cudaSupport`? TBH, I don't like this flag at all, as it makes package maintainers' life pretty complicated, especially for complex packages such as this, and since the cuda support for every package is different and users should read the docs of such packages anyway and not simply count on `config.cudaSupport` to do the job...

Here's a list of scenarios I was able to think about that we should handle:

```
  # config.cudaSupport | packages.GPU | extraCmakeFlags.GPU_API || use cudaPackages.backendStdenv? || is valid?
  # ------------------ | ------------ | ----------------------- || ------------------------------- || ---------
  #      false         |    false     |         "opencl"        ||                                 ||
  #      true          |    false     |         "opencl"        ||                                 ||
  #      false         |    true      |         "opencl"        ||                                 ||
  #      true          |    true      |         "opencl"        ||                                 ||
  #      false         |    false     |         "cuda"          ||                                 ||
  #      true          |    false     |         "cuda"          ||                                 ||
  #      false         |    true      |         "cuda"          ||                                 ||
  #      true          |    true      |         "cuda"          ||                                 ||
  #      false         |    false     |         "hip"           ||                                 ||
  #      true          |    false     |         "hip"           ||                                 ||
  #      false         |    true      |         "hip"           ||                                 ||
  #      true          |    true      |         "hip"           ||                                 ||
```

Not only that, there are the cases where `packages.GPU` is not specified, and/or `extraCmakeFlags.GPU_API` is not specified... And, when `opencl` GPU_API is used, you should add the `ocl-icd` package to `buildInputs`, and `cudaPackages.cudatoolkit` when `cuda` GPU_API is used.

I tried for an hour or so to implement a logic that would satisfy everyone and I got lost in the sea of options above. Not only that, I decided to revert the last revision which read the value of `config.cudaSupport`, as I want the expression to be left clean from heuristic decisions that are hard to figure out just from reading the Nix code.

I guess this is the price we pay for the full configurability of `extraCmakeFlags` (and `packages`) and `extraBuildInputs`. I added a `NOTE` comment explaining this decision.

Reply:
I think understand the problem. However, if the config is too complex and if the intention is to configure it externally via `extraCmakeFlags`, I am non sure `overrideStdenv` is even needed here.
I that case you could override it accordingly and have more transparency: 
```
lammps.override {
  stdenv = cudaPackages.backendStdenv;
  extraCmakeFlags.GPU_API = "cuda";
  <+ what else needs to be overriden>
}
```

Reply:
> I think understand the problem. However, if the config is too complex and if the intention is to configure it externally via `extraCmakeFlags`, I am non sure `overrideStdenv` is even needed here. I that case you could override it accordingly and have more transparency:
> 
> ```
> lammps.override {
>   stdenv = cudaPackages.backendStdenv;
>   extraCmakeFlags.GPU_API = "cuda";
>   <+ what else needs to be overriden>
> }
> ```

When you suggested adding an `overrideStdenv`, I thought you wanted to help the following case: A user wishes to build with `-DGPU_API=cuda` (and `-DPKG_GPU=true`), and not use the default `stdenv`, nor `cudaPackages.backendStdenv`, but something like `clangStdenv` (perhaps for testing the build or whatever). Without an `overrideStdenv` flag, this user would have no way to tell the build to use their supplied `stdenv` argument.

The discussion of having an `overrideStdenv` or not doesn't add or subtract from the complexity described in the above table.